### PR TITLE
Add documentation about JSON serialization and deserialization

### PR DESCRIPTION
**Pull request by davydovanton**

### CODE REVIEW

File: src/json/json.cr [hunk 1]
```diff
@@ -1,3 +1,40 @@
...
+#
+# Generating JSON
+#
+# Creating a JSON string for communication or serialization is just as simple.
+#
+# ```
+# require "json"
+#
+# { hello: "goodbye" }.to_json # => "{\"hello\":\"goodbye\"}"
+# ```
```
Comment (jhass) [inline src/json/json.cr]: I feel like we should also reference `json_mapping` here and maybe even say it's the recommended way to parse JSON with Crystal. See https://github.com/manastech/crystal/issues/982#issuecomment-121156428

**Comment by davydovanton** [comment]
@jhass  thank you so much for comment link. I updated documentation (added section about json type).

@jhass, @asterite what do you think?

File: src/json/json.cr [hunk 2]
```diff
@@ -1,3 +1,84 @@
...
+#
+# ```
+# require "json"
+#
+# json = JSON.parse("{\"response\":{\"domain_data\":\"somedata\"}}")
+# typeof(json)
+# # => (Nil | String | Int64 | Float64 | Bool | Hash(String, JSON::Type) | Array(JSON::Type))
+# ```
+#
+# Thats why you're sure about the structure you can cast:
```
Comment (jhass) [inline src/json/json.cr]: I think there's a missing "if" in this sentence.

File: src/json/json.cr [hunk 3]
```diff
@@ -1,3 +1,84 @@
+# JavaScript Object Notation (JSON)
+#
+# JSON (JavaScript Object Notation) is a way to write data in Javascript.
+# Like XML, it allows to encode structured data in a text format that can
+# be easily read by humans Its simple syntax and native compatibility with
```
Comment (strcmp) [inline src/json/json.cr]: missing a period after "humans" :wink:

File: src/json/json.cr [hunk 4]
```diff
@@ -1,3 +1,84 @@
...
+# curly brackets `{ ... }`.
+#
+# Crystal provides a mechanism for encoding & decoding of values to and
+# from JSON via the serialization API.
+#
+# To read more about JSON visit: json.org
+#
+# Parsing JSON
+#
+# To parse a JSON string received by another application or generated within
```
Comment (strcmp) [inline src/json/json.cr]: tbh i'm not sure the origin of the JSON is relevant. "To parse a JSON string, you can do the following" or just "To parse a JSON string:"

File: src/json/json.cr (hunk 2 above)
Comment (strcmp) [inline src/json/json.cr]: yes it reads a little strange, something like "When you're sure about the structure of the JSON, you can explicitly cast it as a Hash or an Array" sounds better to me, and as far as I know, valid JSON will always enclose other objects in either of those.

File: src/json/json.cr [hunk 5]
```diff
@@ -1,3 +1,84 @@
...
+# ```
+# require "json"
+#
+# json = JSON.parse("{\"response\":{\"domain_data\":\"somedata\"}}")
+#
+# casted = json as Hash
+# casted["response"] # => {"domain_data" => "somedata"}
+# ```
+#
+# If you need to serialize specific JSON type you can use `json_mapping`
```
Comment (strcmp) [inline src/json/json.cr]: I don't know if this is the best explanation for json_mapping but i'm not familiar enough to provide better description.

**Comment by asterite** [comment]
@davydovanton Thanks for the docs! But JSON is a beast and has many little details that I wanted to document. I also did some refactors in the code to make it easier to navigate in the docs (moved `json_mapping` to a `JSON::Mapping` module and include that in `Object`, for example). I did browse your pull request for inspiration :-)

**Comment by davydovanton** [comment]
@asterite okay, I'm happy to help :smiley:

**Pull request closed without merging**